### ISSUE TITLE
Invalid length prefix in chunked body raises ValueError

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6030]
|<img alt="meissenPlate's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| meissenPlate reported|
|-|-|
|Trac ID|trac#6030|
|Type|defect|
|Created|2012-09-29 19:42:10Z|

I saw [#6029](https://github.com/twisted/twisted/issues/6029) "Invalid Content-Length header raises ValueError", where a malformed request was generating an exception in twisted.web.http that was propagating all the way into the reactor, and decided to look for similar bugs in the same file. I found one. Maybe we should be on the look out for others?

This unittest, which should go in twisted.web.test.test_http, demonstrates the problem:

```

    def test_malformed_chunkedEncoding(self):
        """
        If a request uses the I{chunked} transfer encoding, the request body is
        decoded accordingly before it is made available on the request. If the
        request is malformed, we should get a 400 error.
        """
        # See test_chunkedEncoding for the correct form of this request.
        httpRequest = '''\
GET / HTTP/1.0
Content-Type: text/plain
Transfer-Encoding: chunked

MALFORMED_LINE_THIS_SHOULD_BE_'6'
Hello,
14
 spam,eggs spam spam
0

'''
        testcase = self
        didRequest = False
        class MyRequest(http.Request):
            def process(self):
                # This request should fail, so this should never be called.
                testcase.didRequest = True

        try:
            channel = self.runRequest(httpRequest, MyRequest, success=False)
        except ValueError:
            self.fail()
        self.assertFalse(didRequest)
        self.assertEqual(
            channel.transport.value(),
            "HTTP/1.1 400 Bad Request\r\n\r\n")

```

Note that the ValueError I'm catching in that unittest is the one raised by _ChunkedTransferDecoder. It's propogating that far. There's a sensible place to catch it in HTTPChannel.rawDataReceived.

Attachments:

* [unittest_showing_bug.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/570/570a3e528618d879ad7ec0179beb7b3dede2d331/306fbd72b41ea87171c01d2fc3a42afb3ed3a1a9.diff) (1396 bytes) - added by meissenPlate on 2012-09-29 19:42:47Z - Unittest that demonstrates the bug.
* [ticket_6030_patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/570/570a3e528618d879ad7ec0179beb7b3dede2d331/4adae4be531b3edd3031c3327109052bc5fd519d.diff) (2455 bytes) - added by meissenPlate on 2012-10-05 21:55:21Z - Includes a unit test, a simple fix, and a comment informing coders of HTTPChannel's responsibilities.
* [ticket_6030_patch_2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/570/570a3e528618d879ad7ec0179beb7b3dede2d331/fad2132d8c428f39a67247b1899b193c43a14eac.diff) (3784 bytes) - added by meissenPlate on 2012-10-10 21:43:09Z - Fixed problems noted in review. Test case. Fix. News file fragment. A comment in HTTPChannel.

<details><summary>Searchable metadata</summary>

```
trac-id__6030 6030
type__defect defect
reporter__meissenPlate meissenPlate
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1348947730000000 1348947730000000
changetime__1351354525000000 1351354525000000
version__None None
owner__therve therve

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="meissenPlate's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|meissenPlate set owner to meissenPlate<br>meissenPlate set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="meissenPlate's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|meissenPlate commented|
|-|-|

Okay, no one seems to have any ideas about this, so I'll go ahead and put up my patch, doing the obvious thing: catch the ValueError in HTTPChannel.rawDataReceived, spit an arrow back down the connection, and drop it.

This patch also adds the same unit test as the first patch, so you can ignore that one now.

This patch also adds a comment inside HTTPChannel:
```
    # This is the first class outside of the core to handle user input. The 
    # puck stops here for HTTP related errors. If this class doesn't handle, 
    # e.g., a malformed request, then the core will be getting a failure and
    # dumping it in the logs.
```
Fair warning for future coders.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to meissenPlate|
|-|-|

Thanks for the patch! Some comments:

 * The except ValueError should be around the real int() call failing, not the plain dataReceived. It could hide other bugs, and we want to be more strict about the test.

 * You can remove the try/except in the test.

 * The `testcase.didRequest = True` in the test is useless as you don't store didRequest on self previously. In those cases I like to append something to a local list instead of storing things on self.

 * You need to a small NEWS fragment to explain the changes.

**Comment by twisted-trac** [comment]
|<img alt="meissenPlate's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|meissenPlate commented|
|-|-|

I've transformed the ValueError into a _MalformedChunkedDataError, so HTTPChannel can catch it specifically.

I've also corrected the problem with test.didRequest. Don't know how I missed that.

NEWS fragment added.

I also removed the try/except like you suggested. That test now ERRORs instead of FAILs if you run it on the unpatched code. If that's what you want then that's fine.

**Comment by twisted-trac** [comment]
|<img alt="meissenPlate's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|meissenPlate removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @therve|
|-|-|

Looks good! Making some small changes and applying.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set status to `closed`|
|-|-|

(In [36276]) Drop the connection and return a 400 when we get an invalid lenght prefix in
chunked requests

Author: meissenPlate
Revier: therve
Fixes: [#6030](https://github.com/twisted/twisted/issues/6030)
